pyrolite/mineral/normative.py

```diff
@@ -137,95 +133,233 @@ def endmember_decompose(
+    nepheline = np.where((D2 != 0) & (D2>=(4*data['albite'])), data['albite'], nepheline)
+    albite = np.where((D2 != 0) & (D2<(4*data['albite'])), data['albite']-(D2/4), data['albite'])
+    albite = np.where((D2 != 0) & (D2>=(4*data['albite'])), 0, albite)
+    D3 = np.where((D2 != 0) & (D2>=(4*data['albite'])), D2-4*(data['albite']),0)
+    data['albite'] = albite
+    data['nepheline'] = nepheline
+
+    # Leucite
+    leucite = np.where((D3 != 0) & (D3<2*data['orthoclase']), D3/2, 0)
+    leucite = np.where((D3 != 0) & (D3>=2*data['orthoclase']), data['orthoclase'], leucite)
+    orthoclase = np.where((D3 != 0) & (D3<2*data['orthoclase']), data['orthoclase']-(D3/2),data['orthoclase'])
+    orthoclase = np.where((D3 != 0) & (D3>=2*data['orthoclase']), 0, orthoclase)
+    D4 = np.where((D3 != 0) & (D3>=2*data['orthoclase']),D3-(2*data['orthoclase']),0)
+    data['leucite'] = leucite
+    data['orthoclase'] = orthoclase
+
+    # Dicalcium silicate/wollastonite
+    data['cs'] = np.where((D4 != 0) & (D4<data['wollastonite']/2), D4, data['wollastonite']/2)
+    data['cs'] = np.where(D4 == 0,0,data['cs'])
+    wollastonite = np.where((D4 != 0) & (D4<data['wollastonite']/2), data['wollastonite']-(2*D4), data['wollastonite'])
+    wollastonite = np.where((D4 != 0) & (D4>=data['wollastonite']/2), 0, wollastonite)
+    D5 = np.where((D4 != 0) & (D4>=data['wollastonite']/2),D4-(data['wollastonite']/2),0)
+    data['wollastonite'] = wollastonite
+
+
+    #Dicalcium silicate/olivine
+    olivine = np.where((D5 != 0) & (D5<data['diopside']), data['olivine']+(D5/2), data['olivine'])
+    olivine = np.where((D5 != 0) & (D5>=data['diopside']), data['olivine']+(data['diopside']/2), olivine)                     
+    cs = np.where((D5 != 0) & (D5<data['diopside']), data['cs']+(D5/2), data['cs'])
+    data['cs'] = np.where((D5 != 0) & (D5>=data['diopside']), data['cs']+(data['diopside']/2), cs)
+    D6 = np.where((D5 != 0) & (D5>=data['diopside']), D5-data['diopside'],0)                           
+    diopside = np.where(D5<data['diopside'], data['diopside']-D5, data['diopside'])
+    data['diopside'] = np.where(D5>=data['diopside'], 0, diopside) 
+    data['olivine'] = olivine
+    data['cs'] = cs
+
+    # Kaliophilite/leucite
+    kaliophilite = np.where((D6 != 0) & (data['leucite']>=(D6/2)), D6/2, 0)
+    data['kaliophilite'] = np.where((D6 != 0) & (data['leucite']<(D6/2)), data['leucite'],kaliophilite)
+    DEFSIO2 = np.where((D6 != 0) & (data['leucite']<(D6/2)), D6-(2*data['kaliophilite']),0)
+    leucite = np.where((D6 != 0) & (data['leucite']>=(D6/2)), data['leucite']-(D6/2),data['leucite'])
+    data['leucite'] = np.where((D6 != 0) & (data['leucite']<(D6/2)), 0,leucite)
+    # Subdivide hy & di into Mg & Fe??
+
+    # Convert normative minerals to % by multiplying by molecular weights
+    # Will need to do Fe minerals separately
+    norm = data[['apatite', 'thenardite', 'ilmenite', 'orthoclase',
```

Comment:
Something like below would work here to achieve the change back to mass units (noting that all of these columns should already be in `norm` if you created it earlier..).
```python
minerals = [
    ("apatite", mineral_mw["apatite"]),
    ("thenardite", pt.formula("Na2SO4").mass),
    ("ilmenite", mineral_mw["ilmenite"]),
    (
        "orthoclcase",
        pt.formula("KAlSi3O8").mass * 2,
    ),  # check how Or is used above..
    ...
]
norm.loc[:, [m[0] for m in minerals]] *= np.array([m[1] for m in minerals])[None, :]
```
Alternatively you could use mineral formulae as column names together with `df.pyrochem.to_mass()`, but I suspect the composite Fe-Mg components could be a bit messy and you'd probably want to change it back to mineral names afterwards anyway.